Timeout's no longer being applied

See https://github.com/oss-review-toolkit/ort/blob/5bd616f052557f20280c3c758dd5acb889f8aba2/plugins/scanners/fossid/src/test/kotlin/FossIdTest.kt#L567-L580

@OliverO2 I wonder if this is related to the thread changes ?

@OliverO2, no need to be familiar with OSS Review Toolkit specifics, I guess. All I can say right now is that in the [PR that upgrades Kotest from 5.6.2 to 5.7.0](https://github.com/oss-review-toolkit/ort/pull/7443), we start to see [java.lang.AssertionError: Expected exception kotlinx.coroutines.TimeoutCancellationException but no exception was thrown.](https://github.com/oss-review-toolkit/ort/actions/runs/6054838796/job/16436599586?pr=7443#step:4:6611).

--- a/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/concurrency/DedicatedThreadCoroutineDispatcherFactory.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/concurrency/DedicatedThreadCoroutineDispatcherFactory.kt
@@ -3,8 +3,9 @@ package io.kotest.engine.concurrency
 import io.kotest.core.concurrency.CoroutineDispatcherFactory
 import io.kotest.core.test.TestCase
 import kotlinx.coroutines.DelicateCoroutinesApi
-import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.asCoroutineDispatcher
 import kotlinx.coroutines.withContext
+import java.util.concurrent.Executors
 
 /**
  * A [CoroutineDispatcherFactory] that creates a dedicated thread for each test case.
@@ -14,7 +15,7 @@ object DedicatedThreadCoroutineDispatcherFactory : CoroutineDispatcherFactory {
 
    @OptIn(DelicateCoroutinesApi::class)
    override suspend fun <T> withDispatcher(testCase: TestCase, f: suspend () -> T): T =
-      newSingleThreadContext("dedicated").use { dispatcher ->
+      Executors.newSingleThreadExecutor().asCoroutineDispatcher().use { dispatcher ->
          withContext(dispatcher) {
             f()
          }

--- a/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/concurrency/FixedThreadCoroutineDispatcherFactory.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmMain/kotlin/io/kotest/engine/concurrency/FixedThreadCoroutineDispatcherFactory.kt
@@ -4,11 +4,11 @@ import io.kotest.core.concurrency.CoroutineDispatcherFactory
 import io.kotest.core.test.TestCase
 import io.kotest.mpp.Logger
 import kotlinx.coroutines.CoroutineDispatcher
-import kotlinx.coroutines.DelicateCoroutinesApi
-import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.asCoroutineDispatcher
 import kotlinx.coroutines.withContext
 import java.io.Closeable
 import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executors
 import java.util.concurrent.atomic.AtomicInteger
 import kotlin.reflect.KClass
 
@@ -36,8 +36,7 @@ class FixedThreadCoroutineDispatcherFactory(
 
    private val logger = Logger(FixedThreadCoroutineDispatcherFactory::class)
 
-   @OptIn(DelicateCoroutinesApi::class)
-   private val dispatcherPool = List(threads) { newSingleThreadContext("fixed-${it + 1}/$threads") }
+   private val dispatcherPool = List(threads) { Executors.newSingleThreadExecutor().asCoroutineDispatcher() }
 
    private val requestCount = AtomicInteger(0)
 

--- a/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/io/kotest/engine/concurrency/TimeoutTest.kt
+++ b/kotest-framework/kotest-framework-engine/src/jvmTest/kotlin/io/kotest/engine/concurrency/TimeoutTest.kt
@@ -1,0 +1,32 @@
+package io.kotest.engine.concurrency
+
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.core.spec.style.FunSpec
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withTimeout
+
+class TimeoutTest : FunSpec({
+   test("detection with blocking job") {
+      shouldThrow<TimeoutCancellationException> {
+         withTimeout(50) {
+            launch {
+               Thread.sleep(100)
+            }
+         }
+         println("no timeout detected")
+      }
+   }
+
+   test("detection with non-blocking job") {
+      shouldThrow<TimeoutCancellationException> {
+         withTimeout(50) {
+            launch {
+               delay(100)
+            }
+         }
+         println("no timeout detected")
+      }
+   }
+})

